pkg/diff/sql_generator.go

```diff
@@ -1506,26 +1613,108 @@ func (csg *checkConstraintSQLGenerator) Delete(con schema.CheckConstraint) ([]St
 		return nil, fmt.Errorf("check constraints that depend on UDFs: %w", ErrNotImplemented)
 	}
 
-	return []Statement{
-		validateConstraintStatement(csg.tableName, schema.EscapeIdentifier(diff.new.Name)),
-	}, nil
+	return stmts, nil
+}
+
+func (*checkConstraintSQLVertexGenerator) GetSQLVertexId(con schema.CheckConstraint) string {
+	return buildVertexId("checkconstraint", con.Name)
+}
+
+func (csg *checkConstraintSQLVertexGenerator) GetAddAlterDependencies(con, _ schema.CheckConstraint) ([]dependency, error) {
+	deps := []dependency{
+		mustRun(csg.GetSQLVertexId(con), diffTypeDelete).before(csg.GetSQLVertexId(con), diffTypeAddAlter),
+	}
+
+	targetColumns, err := getTargetColumns(con, csg.newSchemaColumnsByName)
+	if err != nil {
+		return nil, fmt.Errorf("getting target columns: %w", err)
+	}
+
+	isOnPreExistingColumn := false
+	if len(targetColumns) == 1 {
+		targetColumn := targetColumns[0]
+		if _, ok := csg.addedColumnsByName[targetColumn.Name]; !ok {
+			isOnPreExistingColumn = true
+		}
+	}
+
+	if isOnPreExistingColumn && isValidNotNullCC(con) {
+		// If the NOT NULL check constraint is on a pre-existing column, then we should ensure it is added before
+		// the column.
+		deps = append(deps, mustRun(csg.GetSQLVertexId(con), diffTypeAddAlter).before(buildColumnVertexId(targetColumns[0].Name), diffTypeAddAlter))
+	} else {
+		for _, tc := range targetColumns {
+			deps = append(deps, mustRun(csg.GetSQLVertexId(con), diffTypeAddAlter).after(buildColumnVertexId(tc.Name), diffTypeAddAlter))
+		}
+	}
+	return deps, nil
+}
+
+func (csg *checkConstraintSQLVertexGenerator) GetDeleteDependencies(con schema.CheckConstraint) ([]dependency, error) {
+	var deps []dependency
+
+	targetColumns, err := getTargetColumns(con, csg.oldSchemaColumnsByName)
+	if err != nil {
+		return nil, fmt.Errorf("getting target columns: %w", err)
+	}
+
+	for _, tc := range targetColumns {
+		// This is a weird quirk of our graph system, where if a -> b and b -> c and b does-not-exist, b will be
+		// implicitly created s.t. a -> b -> c (https://github.com/stripe/pg-schema-diff/issues/84)
+		//
+		// In this case, "a" is the deletion of the check constraint, "b" is the deletion of
+		// the column, and "c" is the alter/addition of the column. We do not want this behavior. We only want
+		// a -> b -> c iff the column is being delted.
+		if _, ok := csg.deletedColumnsByName[tc.Name]; ok {
+			deps = append(deps, mustRun(csg.GetSQLVertexId(con), diffTypeDelete).before(buildColumnVertexId(tc.Name), diffTypeDelete))
+		}
+	}
+
+	// If it's a not-null check constraint, we can drop the check constraint whenever convenient, e.g., after
+	// the column has been altered because `NOT NULL` does not depend on the type of the column.
+	// For all other check constraints, they can rely on the type of the column. Thus, we should drop these
+	// check constraint before any columns are altered because the new type might not be compatible with the old
+	// check constraint.
```

Comment:
Is there value to allowing the NN constraints to be dropped whenever? Maybe it'd be best to add the dependency regardless just for consistency among the constraint types?

Reply:
It's critical we drop this last, which the topographical sort does if an obj has no deps because of the scenario where we're deleting a valid not null constraint AND changing a column to NOT NULL.

I imagine the scenario above is pretty common because it would occur for a half-complete migration, where the valid constraint was added but the `SET NOT NULL` never ran